google/stackdriver_monitoring/prometheus_service.py

```diff
@@ -0,0 +1,187 @@
+        A dictionary maping metric names to MetricInfo being built.
+    """
+    # In practice this converts a Spinnaker Timer into either
+    # <name>__count or <name>__totalTime and removes the "statistic" tag.
+    name, tags = spectator_client.normalize_name_and_tags(
+        name, instance, metric_metadata)
+    if tags is None:
+      return  # ignore metrics that had no tags because these are bogus.
+
+    record = InstanceRecord(service,
+                            '{0}:{1}'.format(service_metadata['__host'],
+                                             service_metadata['__port']),
+                            instance)
+
+    name_to_info = service_to_name_to_info.get(service)
+    if name_to_info is None:
+      name_to_info = {}
+      service_to_name_to_info[service] = name_to_info
+
+    tag_names = set([tag['key'] for tag in tags])
+    info = name_to_info.get(name)
+    if info is None:
+      info = MetricInfo(metric_metadata['kind'], tag_names, [record])
+      name_to_info[name] = info
+      return
+
+    info.records.append(record)
+    info.tags.update(tag_names)
+
+    # In practice this converts a Spinnaker Timer into either
+    # <name>__count or <name>__totalTime and removes the "statistic" tag.
+    name, tags = spectator_client.normalize_name_and_tags(
+        name, instance, metric_metadata)
+    if tags is None:
+      return  # ignore metrics that had no tags because these are bogus.
+
+
+  def collect(self):
+    """Implements Prometheus Client interface."""
+    service_to_name_to_info = {}
+
+    service_metric_map = self.__spectator.scan_by_service(
+        self.__service_endpoints)
+    spectator_client.foreach_metric_in_service_map(
+        service_metric_map, self.__collect_instance_info,
+        service_to_name_to_info)
+
+    all_members = []
+    for service, name_to_info in service_to_name_to_info.items():
+      for name, info in name_to_info.items():
+        family = (CounterMetricFamily
+                  if info.kind in ('Counter', 'Timer')
+                  else GaugeMetricFamily)
+
+        member_name = '{service}:{name}'.format(
+            service=service, name=name.replace('.', ':'))
+
+        tags = list(info.tags)
+        all_tags = list(tags)
+        if self.__add_metalabels:
+          all_tags.extend(['job', 'instance'])
+        member = family(member_name, '', labels=all_tags)
+        all_members.append(member)
+
+        for record in info.records:
+          if isinstance(record, dict):
+            print '*** RECORD {0}'.format(record)
+            print '*** INFO {0}'.format(info)
+
+          instance = record.data
+          labels = [''] * len(tags)
+          for elem in instance['tags']:
+            labels[tags.index(elem['key'])] = elem['value']
+          if self.__add_metalabels:
+            labels.append(record.service)
+            labels.append(record.netloc)
+          member.add_metric(labels=labels, value=instance['values'][0]['v'])
```

Comment:
Do we have any control of the keys of the nested `instance` map? `['values'][0]['v']` doesn't seem to convey much information about what is accessed here. Maybe that's the point though, I'm not extremely familiar with this. Just a thought.